nodejs/wrapper.go

```diff
@@ -0,0 +1,299 @@
+/*
+Copyright 2020 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// A wrapper for node executables to support debugging of application scripts.
+// Many NodeJS applications use NodeJS-based launch tools (e.g., npm,
+// nodemon), and often use several in combination.  This makes it very
+// difficult to start debugging the application as `--inspect`s are usually
+// intercepted by one of the launch tools.  When executing a `node_modules`
+// script, this wrapper strips out and propagates `--inspect`-like arguments
+// via `NODE_DEBUG`.  When executing an app script, this wrapper then inlines
+// the `NODE_DEBUG` when found.
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	shell "github.com/kballard/go-shellquote"
+	"github.com/sirupsen/logrus"
+)
+
+// nodeContext allows manipulating the launch context for node.
+type nodeContext struct {
+	program string
+	args    []string
+	env     map[string]string
+}
+
+func main() {
+	logrus.SetLevel(logrusLevel())
+	logrus.Debugf("Launched: %v", os.Args)
+
+	env := envToMap(os.Environ())
+	// suppress npm warnings when node on PATH isn't the node used for npm
+	env["npm_config_scripts_prepend_node_path"] = "false"
+	nc := nodeContext{program: os.Args[0], args: os.Args[1:], env: env}
+	if err := run(&nc, os.Stdin, os.Stdout, os.Stderr); err != nil {
+		logrus.Fatal(err)
+	}
+}
+
+func logrusLevel() logrus.Level {
+	switch os.Getenv("WRAPPER_VERBOSE") {
+	case "trace":
+		return logrus.TraceLevel
+	case "debug":
+		return logrus.DebugLevel
+	case "info":
+		return logrus.InfoLevel
+	case "warn":
+		return logrus.WarnLevel
+	case "error":
+		return logrus.ErrorLevel
+
+	default:
+		return logrus.WarnLevel
+	}
+}
+
+func run(nc *nodeContext, stdin io.Reader, stdout, stderr io.Writer) error {
+	if !nc.unwrap() {
+		return fmt.Errorf("unwrap could not find actual executable")
+	}
+	logrus.Debugf("unwrapped: %s\n", nc.program)
+
+	// use an absolute path in case we're being run within a node_modules directory
+	script := findScript(nc.args)
+	if abs, err := filepath.Abs(script); err == nil {
```

Comment:
What about when there's an error?

Reply:
Hmm, I was thinking of just treating it as a file name but maybe it's better to skip any processing and just immediately hand off to the unwrapped executable.

Reply:
I think `filepath.Abs()` only errors if it can't determine the current directory